### PR TITLE
Unirep update

### LINKED ISSUE #62: Refactor to use on chain trees

### PR DESCRIPTION
**Pull request by vimwitch**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Closes #62 

## What is the new behavior?

Misc fixes integrating the changes to unirep core packages.

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

## Other information

Depends on
- unirep/unirep#92
- unirep/unirep#91
- unirep/unirep#89
- unirep/unirep#88
- unirep/unirep#109

### CODE REVIEW

**Comment by vimwitch** [comment]
@kittybest This branch has the latest fixes for integrating unirep. To get it working locally pull the `staging` branch in your unirep repo and then build. Then link the packages in unirep-social using `yarn linkUnirep ../path/to/unirep`.

This is an example:

```sh
cd /tmp
git clone -b staging git@github.com:unirep/unirep.git
git clone -b unirep-update git@github.com:unirep/unirep-social.git
cd unirep
yarn && yarn build
yarn workspaces foreach run pre
cd ../unirep-social
yarn
yarn linkUnirep /tmp/unirep
yarn build
```

**Comment by vimwitch** [comment]
This is working pretty well now. Sync percentage, UST, post/comment/vote/airdrop all working.

File: packages/backend/src/utils.ts [hunk 1]
```diff
@@ -14,27 +14,17 @@ const verifyGSTRoot = async (
     epoch: number,
     gstRoot: string
 ): Promise<boolean> => {
-    const exists = await db.findOne('GSTRoot', {
```
Comment (vimwitch) [inline packages/backend/src/utils.ts:17]: Need to actually check these

File: packages/backend/src/utils.ts [hunk 2]
```diff
@@ -223,11 +213,11 @@ const verifyUSTProof = async (
     // check nullifiers
     const exists = await db.findOne('Nullifier', {
         where: {
-            nullifier: results.finalTransitionProof.epochKeyNullifiers,
+            nullifier: results.finalTransitionProof.epkNullifiers,
```
Comment (vimwitch) [inline packages/backend/src/utils.ts:216]: This check can probably be completely removed

File: packages/backend/webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+const path = require('path')
```
Comment (vimwitch) [inline packages/backend/webpack.config.js:1]: Remove this file as it was just for testing

File: packages/core/hardhat.config.ts [hunk 4]
```diff
@@ -10,10 +10,12 @@ const config: HardhatUserConfig = {
...
         },
         optimism: {
-            url: 'https://kovan.optimism.io',
+            url: 'https://goerli.optimism.io',
             accounts: [
                 // 0xeb465b6C56758a1CCff6Fa56aAee190646A597A0
-                '0x18ef552014cb0717769838c7536bc1d3b1c800fe351aa2c38ac093fa4d4eb7d6',
+                // '0x18ef552014cb0717769838c7536bc1d3b1c800fe351aa2c38ac093fa4d4eb7d6',
+                // 0x824B430feb314e0A3F30Dc8A427140071Daeb4E0
+                '0x05d21e78698a4146c449ebf627b1436b23e7a682eef813937c66c8d18b64ff95',
```
Comment (vimwitch) [inline packages/core/hardhat.config.ts:18]: Restore this

File: packages/backend/src/utils.ts [hunk 5]
```diff
@@ -8,33 +8,30 @@ import {
     ProcessAttestationsProof,
 } from '@unirep/contracts'
 import { DB } from 'anondb'
+import { UNIREP, UNIREP_ABI, DEFAULT_ETH_PROVIDER } from './constants'
```
Comment (vivianjeng) [inline packages/backend/src/utils.ts:11]: I think we can use abi from `@unirep/contracts`

**Review by vivianjeng — APPROVED** [review]
LGTM

File: packages/backend/src/utils.ts (hunk 5 above)
Comment (vimwitch) [reply packages/backend/src/utils.ts:11]: Yep, this is changed in #71

**Comment by kittybest** [comment]
The problems I found 
1. the progress bar does not really show the progress (I saw `infinity` or `0%` all the time)
2. cannot show real spent in user page
3. click `activity` in user page will crash
4. frontend sends the user transition request too early
5. didn't get airdrop in the new epoch

I think these can be fixed in another PR, so for now this PR lgtm :)

**Review by kittybest — APPROVED** [review]

**Pull request merged**